### PR TITLE
[Lang] [ir] Support querying external array shape by 'x.shape'

### LINKED ISSUE #1678: [Bug or feature] Can we access the shape of external array?
**Backgrounds**
Since `ti.field` must be declared before any kernel invocation. Making passing data dynamic size impossible.
So I'd like to extent its ability for more flexible data structure.
E.g., [Taichi THREE](https://github.com/taichi-dev/taichi_three) could provide such API:
```py
scene.triangles(arr1, arr2, arr3)  # where arr is numpy array of 3d vectors

# inside taichi_three:
@ti.kernel
def triangles(self, arr1: ti.ext_arr(), ...):  # template id is the (shape, dtype) pair
  ti.static_assert(arr1.shape[1] == 3)
  for i in range(arr1.shape[0]):
    ...
```
However, urrently the `ti.ext_arr()` is very edged and even buggy according [this forum thread](https://forum.taichi.graphics/t/hw0-5-implicit-euler-method-on-mass-spring-system/961).
So I raise about the first step for it:

**Concisely describe the proposed feature**
I'd like to access the shape of an external array in Taichi-scope, e.g.:
```py
import taichi as ti
import numpy as np

@ti.kernel
def my_kernel(x: ti.ext_arr()):
    for i in range(x.shape[0]):
        print(x[i])

x = np.array([2, 3, 4, 5])
my_kernel(x)
```

However, I obtain:
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-ybvnxt8j
[Taichi] <dev mode>, llvm 10.0.0, commit 2be05fc7, python 3.8.3
[Taichi] materializing...
[E 08/10/20 22:06:34.318] [expr.cpp:snode@83] Cannot get snode of non-global variables.


***********************************
* Taichi Compiler Stack Traceback *
***********************************
/tmp/taichi-ybvnxt8j/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
/tmp/taichi-ybvnxt8j/taichi_core.so: taichi::lang::Expr::snode() const
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x4c481f) [0x7fc79aec381f]
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x394272) [0x7fc79ad93272]
/usr/lib/libpython3.8.so.1.0: PyCFunction_Call
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall
/usr/lib/libpython3.8.so.1.0(+0x13e75f) [0x7fc79ef1e75f]
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0(+0x13a378) [0x7fc79ef1a378]
/usr/lib/libpython3.8.so.1.0: _PyObject_GenericGetAttrWithDict
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0(+0x13a378) [0x7fc79ef1a378]
/usr/lib/libpython3.8.so.1.0: _PyObject_GenericGetAttrWithDict
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: PyObject_Call
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x4c6175) [0x7fc79aec5175]
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x4c610f) [0x7fc79aec510f]
/tmp/taichi-ybvnxt8j/taichi_core.so: taichi::lang::Kernel::Kernel(taichi::lang::Program&, std::function<void ()>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool)
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x447f2c) [0x7fc79ae46f2c]
/tmp/taichi-ybvnxt8j/taichi_core.so: taichi::lang::Program::kernel(std::function<void ()> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x456f92) [0x7fc79ae55f92]
/tmp/taichi-ybvnxt8j/taichi_core.so(+0x394272) [0x7fc79ad93272]
/usr/lib/libpython3.8.so.1.0: PyCFunction_Call
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall
/usr/lib/libpython3.8.so.1.0(+0x13e75f) [0x7fc79ef1e75f]
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0(+0x13e442) [0x7fc79ef1e442]
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyObject_FastCallDict
/usr/lib/libpython3.8.so.1.0: _PyObject_Call_Prepend
/usr/lib/libpython3.8.so.1.0(+0x1f5e09) [0x7fc79efd5e09]
/usr/lib/libpython3.8.so.1.0: PyObject_Call
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: PyEval_EvalCode
/usr/lib/libpython3.8.so.1.0(+0x1d8248) [0x7fc79efb8248]
/usr/lib/libpython3.8.so.1.0(+0x1d2483) [0x7fc79efb2483]
/usr/lib/libpython3.8.so.1.0: PyRun_FileExFlags
/usr/lib/libpython3.8.so.1.0: PyRun_SimpleFileExFlags
/usr/lib/libpython3.8.so.1.0: Py_RunMain
/usr/lib/libpython3.8.so.1.0: Py_BytesMain
/usr/lib/libc.so.6: __libc_start_main
python(_start+0x2e) [0x561f5791804e]

Internal Error occurred, check this page for possible solutions:
https://taichi.readthedocs.io/en/stable/install.html#troubleshooting
Traceback (most recent call last):
  File "a.py", line 10, in <module>
    my_kernel(x)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 571, in wrapped
    return primal(*args, **kwargs)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 498, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 374, in materialize
    taichi_kernel = taichi_kernel.define(taichi_ast_generator)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 371, in taichi_ast_generator
    compiled()
  File "a.py", line 6, in my_kernel
    for i in range(x.shape[0]):
  File "/home/bate/Develop/taichi/python/taichi/lang/expr.py", line 140, in shape
    return self.snode.shape
  File "/home/bate/Develop/taichi/python/taichi/lang/expr.py", line 133, in snode
    return SNode(self.ptr.snode())
RuntimeError: [expr.cpp:snode@83] Cannot get snode of non-global variables.
```
instead.

**Describe the solution you'd like (if any)**
It seems that in Taichi-scope, the `x` is not of type `np.array`, but of type `ti.Expr`? So I suggest:
1. Let `x: ti.ext_arr()` to be simply a `np.array`.
2. Do mocks in `ti.subscript(x, indices)` when `x` is detected to be `np.array`, do the original stuffs.
So accessing `shape` or `dtype` would be strightfoward.

**Additional comments**

### PR DESCRIPTION
**Pull request by thinking-tower**
Fixes #1678. cc @archibate.

# Description
Currently, any access to an external array's shape throws an error. For example, (in #1678 as mentioned by @archibate)
```
import taichi as ti
import numpy as np

@ti.kernel
def my_kernel(x: ti.ext_arr()):
    for i in range(x.shape[0]):
        print(x[i])

x = np.array([2, 3, 4, 5])
my_kernel(x)
```
throws `[expr.cpp:snode@83] Cannot get snode of non-global variables`.

This PR provides the functionality to access the shape (per axis) of a `ti.ext_arr()`.

# Notes
1. This won't work however:
```
@ti.kernel
def my_kernel(x: ti.ext_arr()):
    print(x.shape) # ValueError: Bad constant scalar expression: <class 'list'>

x = np.array([2, 3, 4, 5])
my_kernel(x)
````
2. I'm imitating how shape is obtained in `snode.py`:
https://github.com/taichi-dev/taichi/blob/f1bde29bd7ba32712b59426c19b6877cf5e7f0f8/python/taichi/lang/snode.py#L80-L91

### CODE REVIEW

**Comment by thinking-tower** [comment]
@archibate What should I put for `ExternalTensorShapeAlongAxisExpression::serialize() override`?

File: tests/python/test_get_external_tensor_shape.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import numpy as np
+
+
+@ti.all_archs
+def test_get_external_tensor_shape_access():
+    @ti.kernel
+    def func(x: ti.ext_arr(), index: ti.template()) -> ti.i32:
+        return x.shape[index]
+
+    size = np.random.randint(low=1, high=10, size=(8, ))
```
Comment (yuanming-hu) [inline tests/python/test_get_external_tensor_shape.py:11]: Let's use a deterministic `size` here (e.g. 7). If a unit test fails, we want it to fail deterministically :-)

File: tests/python/test_get_external_tensor_shape.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+def test_get_external_tensor_shape_sum():
+    @ti.kernel
+    def func(x: ti.ext_arr()) -> ti.i32:
+        y = 0
+        for i in range(x.shape[0]):
+            for j in range(x.shape[1]):
+                y += x[i, j]
+        return y
+
+    size = np.random.randint(low=1, high=10, size=(2, ))
```
Comment (yuanming-hu) [inline tests/python/test_get_external_tensor_shape.py:29]: Again, a deterministic shape would work better :-)

File: tests/python/test_get_external_tensor_shape.py (hunk 1 above)
Comment (archibate) [reply tests/python/test_get_external_tensor_shape.py:11]: Yes, please consider make use of `@pytest.mark.parametrize` and use a list of representitive shapes as parameter, e.g.:
- 3
- (3, )
- (5, 6)
- (3, 1, 5)

**Review by k-ye — APPROVED** [review]
Confirmed that this worked on Metal, great work!!

File: tests/python/test_get_external_tensor_shape.py (hunk 1 above)
Comment (thinking-tower) [reply tests/python/test_get_external_tensor_shape.py]: @yuanming-hu, @archibate I've now made the arrays fixed size and initialised with fixed values relative to the minimum number of bits an `int` can have as specified [here](https://en.cppreference.com/w/cpp/language/types). 

Note I've squashed my commits in this branch however the original version can still be seen [here](https://github.com/thinking-tower/taichi/pull/2/commits).

Comment (thinking-tower) [reply tests/python/test_get_external_tensor_shape.py]: In hindsight, I realized that it's possible to run out of CUDA memory if the arrays are too big. I've reduced the size of the tests and observed that the test assume that you have a NVIDIA GPU with > `2.5`GB. In #356, it's mentioned that taichi won't support "pre-Pascal GPUs". In the [wikipedia page](https://en.wikipedia.org/wiki/GeForce_900_series#GeForce_900_(9xx)_series), there is a GPU with memory `1024 MB`. Would this be an issue?

**Review by archibate — APPROVED** [review]
Nice work! Impressive quality for first-time contributors!

File: tests/python/test_get_external_tensor_shape.py (hunk 1 above)
Comment (yuanming-hu) [reply tests/python/test_get_external_tensor_shape.py]: We actually do support pre-Pascal GPUs as long as you set `TI_USE_UNIFIED_MEMORY=0`. For test cases, we simply always limited memory usage < 256MB :-) Your current implementation looks good.

**Pull request merged**